net/golioth/coap_req.c

```diff
@@ -0,0 +1,709 @@
+{
+	return (v1 < v2 && v2 - v1 < (1 << 23)) ||
+		(v1 > v2 && v1 - v2 > (1 << 23));
+}
+
+static int golioth_coap_req_reply_handler(struct golioth_coap_req *req,
+					  const struct coap_packet *response)
+{
+	uint16_t payload_len;
+	uint8_t code;
+	const uint8_t *payload;
+	int block2;
+	int err;
+
+	code = coap_header_get_code(response);
+
+	LOG_DBG("CoAP response code: 0x%x (class %u detail %u)",
+		(unsigned int)code, (unsigned int)(code >> 5), (unsigned int)(code & 0x1f));
+
+	err = golioth_coap_code_to_posix(code);
+	if (err) {
+		struct golioth_req_rsp rsp = {
+			.user_data = req->user_data,
+			.err = err,
+		};
+
+		(void)req->cb(&rsp);
+
+		LOG_INF("cancel and free req: %p", req);
+
+		goto cancel_and_free;
+	}
+
+	payload = coap_packet_get_payload(response, &payload_len);
+
+	block2 = coap_get_option_int(response, COAP_OPTION_BLOCK2);
+	if (block2 != -ENOENT) {
+		size_t new_offset;
+		size_t cur_offset;
+
+		err = coap_update_from_block(response, &req->block_ctx);
+		if (err) {
+			struct golioth_req_rsp rsp = {
+				.user_data = req->user_data,
+				.err = -EBADMSG,
+			};
+
+			LOG_ERR("Failed to parse get response: %d", err);
+
+			(void)req->cb(&rsp);
+
+			err = -EBADMSG;
+			goto cancel_and_free;
+		}
+
+		cur_offset = req->block_ctx.current;
+
+		new_offset = coap_next_block_for_option(response, &req->block_ctx,
+							COAP_OPTION_BLOCK2);
+		if (new_offset < 0) {
+			struct golioth_req_rsp rsp = {
+				.user_data = req->user_data,
+				.err = -EBADMSG,
+			};
+
+			LOG_ERR("Failed to move to next block: %d", new_offset);
+
+			(void)req->cb(&rsp);
+
+			err = -EBADMSG;
+			goto cancel_and_free;
+		} else if (new_offset == 0) {
+			struct golioth_req_rsp rsp = {
+				.data = payload,
+				.len = payload_len,
+				.off = cur_offset,
+
+				.total = req->block_ctx.total_size,
+
+				.get_next = NULL,
+				.get_next_data = NULL,
+
+				.user_data = req->user_data,
+			};
+
+			LOG_DBG("Blockwise transfer is finished!");
+
+			(void)req->cb(&rsp);
+
+			goto cancel_and_free;
+		} else {
+			struct golioth_req_rsp rsp = {
+				.data = payload,
+				.len = payload_len,
+				.off = cur_offset,
+
+				.total = req->block_ctx.total_size,
+
+				.get_next = golioth_coap_req_next_block,
+				.get_next_data = req,
+
+				.user_data = req->user_data,
+
+				.err = req->is_observe ? -EMSGSIZE : 0,
+			};
+
+			err = req->cb(&rsp);
+			if (err) {
+				LOG_WRN("Received error (%d) from callback, cancelling", err);
+				goto cancel_and_free;
+			}
+
+			if (req->is_observe) {
+				LOG_ERR("TODO: blockwise observe is not supported");
+				err = -ENOTSUP;
+				goto cancel_and_free;
+			}
+
+			return 0;
+		}
+	} else {
+		struct golioth_req_rsp rsp = {
+			.data = payload,
+			.len = payload_len,
+			.off = 0,
+
+			/* Is it the same as 'req->block_ctx.total_size' ? */
+			.total = payload_len,
+
+			.get_next = NULL,
+			.get_next_data = NULL,
+
+			.user_data = req->user_data,
+		};
+
+		(void)req->cb(&rsp);
+
+		goto cancel_and_free;
```

Comment:
In the blockwise case, after calling the callback, the return value is checked. Any reason we're not checking it here?

Reply:
Actually only in case of blockwise transfer this return value is checked and if it is non-zero, then blockwise transfer is not continued. This is a special case and subject to refinement in the future. The idea was to use (just in blockwise transfer case) return code to signal whether application want to continue receiving next blocks of data or stop (e.g. when downloaded content does not fit into preallocated memory or in case of any other error).

This is part of asynchronous API discussion/improvement. For now this API is marked in Doxygen comments as "experimental", mainly because of the details on how to handle blockwise transfer. Also 2 issues were added to Jira, which described the direction of (possible) further improvements.